[RISCV] Handle more (add x, C) -> (sub x, -C) cases

pfusik opened the pull request:
This is a follow-up to #137309, adding:
- multi-use of the constant with different adds
- vectors (vadd.vx -> vsub.vx)

pfusik commented on llvm/lib/Target/RISCV/RISCVInstrInfoVSDPatterns.td:
```diff
@@ -907,6 +907,19 @@ foreach vti = AllIntegerVectors in {
+  let Predicates = [HasVInstructionsI64] in {
+    def : Pat<(add (vti.Vector vti.RegClass:$rs1),
+                   (vti.Vector (SplatPat (i64 negImm:$rs2)))),
+//                 (riscv_vmv_v_x_vl undef, negImm:$rs2, srcvalue)),
```
This doesn't match the `vadd_vx_imm64_to_sub` test yet (neither `SplatPat` nor `riscv_vmv_v_x_vl`).
I'd appreciate hints for how to debug. Or maybe you can see why it isn't working?
Perhaps I need to add `SplatPat_simmXXX` ? If so, why?

topperc replied:
I think its because ComplexPatterns can't nest. SplatPat and negImm are both ComplexPatterns.

pfusik replied:
Resolved by adding a combined ComplexPattern.
Pity there's no TableGen diagnostic for nested ComplexPatterns.

pfusik committed:
[RISCV] Handle more (add x, C) -> (sub x, -C) cases

This is a follow-up to #137309, adding:
- multi-use of the constant with different adds
- vectors (vadd.vx -> vsub.vx)

topperc reviewed (approved):
LGTM

pfusik commented:
Tests merged into main branch as e618a79711a7aefa442f8a417aceeac2ceadbcdd. PR rebased.
